docs/hoon/concepts.md

```diff
@@ -82,16 +77,53 @@ it's still fun and useful to [learn more](../../nock).
 
 ## Compilation concepts
 
+### <a name="hoon">`hoon`</a> (AST node)
+
+A [`hoon`](../reference) is the result of parsing a Hoon source 
+expression into an AST node. Because every Hoon programs is, 
+in its entirety, a single expression of Hoon, the result of parsing 
+a whole Hoon program into an AST is a single `hoon`.
+
+As the noun that the parser produces, a hoon is a tagged union of
+the form `[tag data]`, where the tag is a constant such as `%brts`,
+which is matched up with the appropriate type of data (often more 
+`hoon`s).  For example, the expression `:-(p q)`, once parsed into 
+an AST, becomes the tagged union:
+
+```
+[%clhp p=hoon q=hoon]
```

Comment:
This is a bit confusing. The *type* of the parsed expression is a tagged union, which includes all the possible tags, not just the `%clhp`. You'd need a `?-` or equivalent to get the type to be expressed this way. Maybe mention that `++type` is a tagged union, and the `%clhp` subtype looks like this. The *value* of the parsed expression, as you explain below, has that form.

Reply:
I think I'm not understanding you here.  I'm saying that `[%clhp p=hoon q=hoon]` is a tagged union in the same sense that `17` is an `@` or an `atom`.  Yes, `@` is a mold which defines the type `atom`, but it's pretty clear that I don't mean `17` is a mold or a type.  In fact, it seems clear to me that `++hoon` isn't a tagged union, because it's a mold!  In the same way, `@` itself isn't an `atom`...

What am I missing?

Reply:
The term "tagged union" usually refers to a type, not a value. So `++hoon` is a mold that normalizes to a tagged union type. When you say that `[%clhp p=hoon q=hoon]` is a tagged union, that is confusing to me because it's really just one subtype of the full tagged union type. To me, the full tagged union would be the `$%  [%clhp ...]  [%cnts ...] ... ==`.

Also, saying `[%clhp p=hoon q=hoon]` is a tagged union is different from saying that `17` is an `atom` because it's not an instance of the type the way `17` is an instance of `atom`. An instance of the type would look more like `[%clhp p=*hoon q=*hoon]`, or your later example with the `%sand`s.

Does that make sense?

Reply:
There is something funky going on with language here.  I hope it's okay if I probe a bit to get a better sense of what's going wrong.  Pedantry ahead...

>  The term "tagged union" usually refers to a type, not a value.

Agreed!  In the same way, "integer" refers to a type, not a value; and we still say "15 is an integer".  This usage clearly means "15 is an instance of the type, integers".  Even though 15 is of a sub-type (say, the odd integers, or integers divisible by 5, etc.) of the full type of integers.

So it still seems okay to say (with a little imprecision):

"`[%clhp p=hoon q=hoon]` is a tagged union."

Oh wait, no it doesn't.  (I am/was wrong!)  Which tagged union type?  There are lots!  `++type`, `++hoon`, `++card`... these all (seem to) define distinct tagged union types.

The problem is that "tagged union" doesn't seem to be (simply) a type.  It's a *type* of type.  (Does this seem right?)

Accordingly, I *don't* think we should say that *this* is a tagged union either: `$% [%clhp ...] [%cnts ...] ... ==`.  That's because this is a mold, and a mold is a gate, not a type.  This mold *defines* the type for parsed Hoon ASTs, without itself *being* that type.  (`@` isn't a type, it's a mold! Speaking loosely sinks ships...? or it lets you avoid being an annoying pedant, or something....)

Here's what's confusing me though.  All tags look like this: `%foo`.  First a % and then some text (and/or integers).  "$foo" isn't a tag!  So the mold for a tagged union doesn't have tags.  The type defined by the `++hoon` mold is a set that contains the union of the tag-types `[%brts (something)]`, `[%clhp (something)]`, etc.

This is one case where I can't blame Tlon for the confusing vocabulary.  A "tagged union" is a union that isn't itself tagged, but which contains pairs in which the LHS is tagged.

Does this all seem right?

In this case, what's the handy-dandy way to refer to an instance of the `++hoon` tagged union type?  In gloriously ambiguous (but probably totally safe) fashion, do we call it a `hoon`?

Thanks in advance for providing me a CS education without compensation.

Reply:
Despite my disclaimer about pedantry, the above post is rather sloppy... if it's not obvious what is meant, don't bother slogging through it, just yell and I'll try again...

Reply:
A type can be a tagged-union (ie, tagged-union is a sub-category of the category type). A value can be an instance of a type.

We say that 15 *is* an integer, but we don't say that 15 *is* a numeric type -- the relationship is further attenuated. Similarly, a given noun may be an instance of a type that is a tagged-union, but saying the noun *is* a tagged-union is overly dis-intermediated.

Reply:
That sounds right Joe.